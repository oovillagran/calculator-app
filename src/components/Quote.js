import { useState, useEffect } from 'react';

function Quote() {
  const [data, setData] = useState([]);
  const [isLoading, setIsLoading] = useState(false);
  const [hasError, setHasError] = useState(false);
  const category = 'faith';
  const url = 'https://api.api-ninjas.com/v1/quotes?category=';
  const URL = url + category;

  useEffect(() => {
    const fetchData = async () => {
      setIsLoading(true);
      try {
        const response = await fetch(URL, {
          headers: { 'X-Api-Key': 'VCuPtFKrZqieJvVJIYMd9w==CYE6lboEGOeZOd8d' },
        });
        const json = await response.json();
        setData(json);
      } catch (error) {
        setHasError(true);
      }
      setIsLoading(false);
    };
    fetchData();
  }, [URL, setIsLoading]);

  if (isLoading) {
    return (
      <div className="loading">
        <div className="lds-ring">
          <div />
          <div />
          <div />
          <div />
        </div>
      </div>
    );
  }

  if (hasError) return <div className="error">Something went wrong!</div>;

  return (
    <div className="quote-section">
      <div className="quote-container">
        <h2>Quote:</h2>
        {data.map((item) => (
          <p className="quote-text" key={item.quote}>{item.quote}</p>
        ))}
        {data.map((item) => (
          <h3 className="quote-author" key={item.author}>{item.author}</h3>
        ))}
      </div>
    </div>
  );
}

export default Quote;
