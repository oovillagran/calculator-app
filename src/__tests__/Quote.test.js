import { render, screen } from '@testing-library/react';
import Quote from '../components/Quote';
import '@testing-library/jest-dom/extend-expect';

describe('Quote', () => {
  test('renders quote and author', async () => {
    const mockData = [
      {
        author: 'John Doe',
        quote: 'Lorem ipsum dolor sit amet',
      },
    ];
    jest.spyOn(global, 'fetch').mockImplementation(() => Promise.resolve({ json: () => Promise.resolve(mockData) }));

    render(<Quote />);
    const quoteText = await screen.findByText(/Lorem ipsum dolor sit amet/i);
    const quoteAuthor = await screen.findByText(/John Doe/i);

    expect(quoteText).toBeInTheDocument();
    expect(quoteAuthor).toBeInTheDocument();

    global.fetch.mockRestore();
  });

  test('renders loading spinner while fetching data', async () => {
    jest.spyOn(global, 'fetch').mockImplementation(() => new Promise(() => {}));

    render(<Quote />);
    const loadingSpinner = await screen.findByRole('alert');

    expect(loadingSpinner).toBeInTheDocument();

    global.fetch.mockRestore();
  });

  test('renders error message when data fetching fails', async () => {
    jest.spyOn(global, 'fetch').mockImplementation(() => Promise.reject(new Error('Fetch failed')));

    render(<Quote />);
    const errorMessage = await screen.findByText(/Something went wrong!/i);

    expect(errorMessage).toBeInTheDocument();

    global.fetch.mockRestore();
  });

  test('renders the quote page', () => {
    const { container } = render(<Quote />);
    expect(container).toMatchSnapshot();
  })
});
