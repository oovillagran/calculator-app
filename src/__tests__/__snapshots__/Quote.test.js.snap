// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quote renders the quote page 1`] = `
<div>
  <div
    class="loading"
    role="alert"
  >
    <div
      class="lds-ring"
    >
      <div />
      <div />
      <div />
      <div />
    </div>
  </div>
</div>
`;
